core/src/main/java/io/questdb/network/IODispatcherLinux.java

```diff
@@ -45,66 +53,221 @@ public void close() {
+    }
+
+    private long nextEventId() {
+        return (idSeq++ << 1) + 1;
+    }
+
+    private long nextOpId() {
+        return idSeq++ << 1;
+    }
+
+    private void processIdleConnections(long idleTimestamp) {
         int count = 0;
-        for (int i = 0, n = pending.size(); i < n && pending.get(i, M_TIMESTAMP) < deadline; i++, count++) {
-            doDisconnect(pending.get(i), DISCONNECT_SRC_IDLE);
+        for (int i = 0, n = pending.size(); i < n && pending.get(i, OPM_TIMESTAMP) < idleTimestamp; i++, count++) {
+            doDisconnect(pending.get(i), pending.get(i, OPM_ID), DISCONNECT_SRC_IDLE);
         }
         pending.zapTop(count);
     }
 
     private boolean processRegistrations(long timestamp) {
+        boolean useful = false;
         long cursor;
-        int offset = 0;
         while ((cursor = interestSubSeq.next()) > -1) {
-            IOEvent<C> evt = interestQueue.get(cursor);
-            C context = evt.context;
-            int operation = evt.operation;
+            IOEvent<C> event = interestQueue.get(cursor);
+            C context = event.context;
+            int requestedOperation = event.operation;
             interestSubSeq.done(cursor);
 
-            int fd = (int) context.getFd();
-            final long id = fdid++;
+            useful = true;
+            final long opId = nextOpId();
+            final int fd = context.getFd();
+            int operation = requestedOperation;
+            LOG.debug().$("processing registration [fd=").$(fd)
+                    .$(", op=").$(operation)
+                    .$(", id=").$(opId).I$();
+
+            final SuspendEvent suspendEvent = context.getSuspendEvent();
+            if (suspendEvent != null) {
+                // if the operation was suspended, we request a read to be able to detect a client disconnect 
+                operation = IOOperation.READ;
+            }
+
+            int opRow = pending.addRow();
+            pending.set(opRow, OPM_TIMESTAMP, timestamp);
+            pending.set(opRow, OPM_FD, fd);
+            pending.set(opRow, OPM_ID, opId);
+            pending.set(opRow, OPM_OPERATION, requestedOperation);
+            pending.set(opRow, context);
+
             // we re-arm epoll globally, in that even when we disconnect
             // because we have to remove FD from epoll
-            LOG.debug().$("registered [fd=").$(fd).$(", op=").$(operation).$(", id=").$(id).$(']').$();
-            epoll.setOffset(offset);
-            if (epoll.control(fd, id, EpollAccessor.EPOLL_CTL_MOD, operation == IOOperation.READ ? EpollAccessor.EPOLLIN : EpollAccessor.EPOLLOUT) < 0) {
-                System.out.println("oops2: " + nf.errno());
+            final int epollOp = operation == IOOperation.READ ? EpollAccessor.EPOLLIN : EpollAccessor.EPOLLOUT;
+            if (epoll.control(fd, opId, EpollAccessor.EPOLL_CTL_MOD, epollOp) < 0) {
+                LOG.error().$("epoll_ctl modify operation failure [id=").$(opId)
```

Comment:
should be critical

Reply:
Changed in bbec37baf832db9a846d04e59ab124f8bb84eadc